cudamapper/src/overlapper.cpp

```diff
@@ -38,8 +45,118 @@ void Overlapper::update_read_names(std::vector<Overlap>& overlaps,
     }
 }
 
-void Overlapper::print_paf(const std::vector<Overlap>& overlaps)
+void Overlapper::align_overlaps(std::vector<Overlap>& overlaps,
+                                const claragenomics::io::FastaParser& query_parser,
+                                const claragenomics::io::FastaParser& target_parser,
+                                int32_t num_batches,
+                                std::vector<std::string>& cigar)
 {
+    // Calculate max target/query size in overlaps
+    int32_t max_query_size  = 0;
+    int32_t max_target_size = 0;
+    for (const auto& overlap : overlaps)
+    {
+        int32_t query_overlap_size  = overlap.query_end_position_in_read_ - overlap.query_start_position_in_read_;
+        int32_t target_overlap_size = overlap.target_end_position_in_read_ - overlap.target_start_position_in_read_;
+        if (query_overlap_size > max_query_size)
+            max_query_size = query_overlap_size;
+        if (target_overlap_size > max_target_size)
+            max_target_size = target_overlap_size;
+    }
+
+    // Heuristically calculate max alignments possible with available memory based on
+    // empirical measurements of memory needed for alignment per base.
+    const float memory_per_base = 0.03f; // Estimation of space per base in bytes for alignment
+    float memory_per_alignment  = memory_per_base * max_query_size * max_target_size;
+    size_t free, total;
+    CGA_CU_CHECK_ERR(cudaMemGetInfo(&free, &total));
+    const size_t max_alignments = (free * 85 / 100) / memory_per_alignment; // Using 85% of available memory
+    int32_t batch_size          = std::min(get_size<int32_t>(overlaps), static_cast<int32_t>(max_alignments)) / num_batches;
+    std::cerr << "Aligning " << overlaps.size() << " overlaps (" << max_query_size << "x" << max_target_size << ") with batch size " << batch_size << std::endl;
+
+    int32_t overlap_idx = 0;
+    std::mutex overlap_idx_mtx;
+
+    std::vector<std::future<void>> align_futures;
+    for (int32_t t = 0; t < num_batches; t++)
```

Comment:
I think I get it now, it is basically to have multiple cudaaligners per device to hide the host-device copies, right? Wouldn't `num_alignment_engines` or similar be a better name then and shouldn't we provide a good default then?

Reply:
>  it is basically to have multiple cudaaligners per device to hide the host-device copies, right? 
yep that's right

I can rename the variable to `num_alignment_engines` and add a suggested engine count to the help message. it would vary based on how many threads are there in the system, and the size of typical alignments in the batch (but usually between 2-4 is good per device)

Reply:
done